### PR TITLE
threading/Queue: lock mutex before notifying the condition_variable

### PR DESCRIPTION
**Pull request by MaxKellermann**
Locking the associated mutex is not strictly mandatory, but not doing
so can easily create race conditions and lockups.

### CODE REVIEW

**Comment by jsiwek** [comment]
Thanks, but doesn't look necessary.  Merging shouldn't hurt anything, but also don't see it helping anything either -- notifying under locked mutex isn't strictly necessary like you say and usually described as pessimistic.  Essentially, `Wakeup()` is manually inducing a spurious wakeup which the notified-threads have to be designed to accommodate anyway (and looks like they are).

If you had more explanation or example of a real issue caused by lack of locking here, let me know.

**Comment by MaxKellermann** [comment]
> usually described as pessimistic.

Being extremely pessimstic is the only way to write reliable multi-threaded software. The worst case *will* occur, sooner or later, and will produce a less-than-perfectly-happy Zeek user.

> Essentially, `Wakeup()` is manually inducing a spurious wakeup which the notified-threads have to be designed to accommodate anyway (and looks like they are).

I think you misunderstand the problem of notifying without the lock. The problem is not spurious wakeups, but the *opposite*: **missed wakeups**!

If you are not holding the lock while sending the notification, you're not synchronized at all with the target thread. The target thread may be in any random state at this point, for example it may have just entered `pthread_cond_wait()` but has not yet emitted the system call. This way, the notification is ignored, nobody is woken up - and at the same time, the target thread enters the kernel, and waits forever, because nobody will ever wake it up.

A `pthread_cond_t` has no memory of past, missed notifications. If somebody calls `pthread_cond_signal()` and nobody is waiting, nothing happens, and the next `pthread_cond_wait()` call will block forever.

Unless, for some reason, another thread calls `pthread_cond_signal()` again. If you can guarantee that, then locking is not necessary (still a good idea), but is that the case with Zeek?

**Comment by MaxKellermann** [comment]
Just in case you read one of those optimization tricks on the net where they unlock the mutex before signalling the condition variable: yes, that is an optimization to avoid unnecessary contention on the target thread because the notifying thread is still holding the mutex.

**But** it only works in some cases. It works in the most common cases. But it doesn't work with how Zeek does thread synchronization. That's because Zeek's threading code is very clumsy. Apparently, there were deadlock problems long ago which were solved by limiting the wait time to 5 seconds, so Zeek would always recover from deadlock bugs like the one this PR fixes after at least 5 seconds:
https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/Queue.h#L160
That's bad coding style in my opinion, because it sidesteps from fixing the bug, and applies a kludge instead.
Using `wait_for()` is almost always a sign for a deadlock bug swept under the carpet...

This code comment (added by 87e10b5f97a897f8c5fac2f983379a8c8966dcae) makes me more confident that this is the wrong way to do it:
https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/MsgThread.cc#L270-L273

In my opinion, a good program should never depend on timeouts for cancellation (or shutdown); cancellation should always be immediate. If that can't be done, you have made a mistake somewhere. Immediate cancellation is not an optimization - that's how it should always work.

**Comment by jsiwek** [comment]
Thanks for the persistence.  I do now see a potential race condition here as legit, but still wonder if this PR is a "complete" solution.  More on that at the end.

> Just in case you read one of those optimization tricks on the net

You know, my intuition/memory was initially going the way this PR proposed, but I decided to read cppreference which suggests a bit too strongly that notifying under lock isn't needed:

https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all
https://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

I guess if you read carefully, you realize it's technically true *if you know the target thread is already waiting*.  Your description frames things better:

> If you are not holding the lock while sending the notification, you're not synchronized at all with the target thread. The target thread may be in any random state at this point, for example it may have just entered `pthread_cond_wait()` but has not yet emitted the system call. This way, the notification is ignored, nobody is woken up - and at the same time, the target thread enters the kernel, and waits forever, because nobody will ever wake it up.

Agreed, but can we take this thought even further?  Here's the more complete context of the `wait_for()`:

https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/Queue.h#L152-L162

What if the target thread is right at the start of `Get()`, before acquiring the lock?  If `WakeUp()` happens completely when the target thread is still at that point of `Get()`, it can still fall through and hit the `wait_for()` because `WakeUp()` (even if it held the lock while notifying) did nothing to alter the conditions under which the target thread enters `wait_for()`.

Does that sound right to you?  If so, think the PR is just pushing the race to a different place.  A suggestion could be to do 2 things:

* acquire lock from `WakeUp()` (as the PR already suggests)
* add an additional "woke" state to threads that gets set (permanently) from `WakeUp()` (under lock) and checked from `Get()` before entering `wait_for()`

Also agree about potential robustness issues you point out in the general sense, but if we want to do something about this particular race, that's my quick suggestion.

**Comment by MaxKellermann** [comment]
> Does that sound right to you?

Yes.

> If so, think the PR is just pushing the race to a different place.

Yes and no. This PR fixes the bug it is supposed to fix, but does not fix all of Zeek's synchronization bugs, just as you mentioned.

A wait must always wait for an "external condition" (a "predicate"). Prior to invoking the wait, the program must always check if this condition is already true, the program must not wait, but act according to this condition. Wait only if we have something to wait for, but if we have something to do, don't wait, don't yield the CPU. That way, the race you described is not possible.

This condition can be the kill flag, and indeed it is already being checked: `reader->Killed()`. There were two problems with this:
1. the kill flag (i.e. `BasicThread::killed`) is not thread-safe; it is read and written without holding any lock, and it's not an atomic variable
2. there is a race condition between checking this kill flag and the wait

Number 2 is fixed by this PR.
Number 1 is, of course, still buggy, and not addressed by this PR - because it's a different kind of problem.

btw. the most elegant way to wait for this predicate is to use the `condition_variable::wait()` overload which takes a predicate (https://en.cppreference.com/w/cpp/thread/condition_variable/wait). Of course, the predicate must be thread-safe, and the easiest way to do that is to protect all underlying variables with the mutex which is passed to `wait()`. (Don't use atomics. You need a mutex anyway, so doing atomics only adds more overhead.)

**Comment by jsiwek** [comment]
> This condition can be the kill flag, and indeed it is already being checked: `reader->Killed()`. There were two problems with this:
> 
> 1. the kill flag (i.e. `BasicThread::killed`) is not thread-safe; it is read and written without holding any lock, and it's not an atomic variable
> 2. there is a race condition between checking this kill flag and the wait

Yeah, the `kill` flag *can* be the thing to check, but I had doubts about it actually working because of this code path:

https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/MsgThread.cc#L245-L246

So if the thread isn't killed, wake up (and don't change the killed state), but also if the thread isn't killed, it's allowed to wait.

> Number 2 is fixed by this PR.

Still don't think so unless I read the above logic incorrectly.  That's why I was suggesting a separate `woke` state, since `WakeUp` can be called when `killed` is false and does nothing to try to change that state in order to prevent the target thread from waiting.

> Number 1 is, of course, still buggy, and not addressed by this PR - because it's a different kind of problem.

Yeah, unfortunately likely true/separable.

**Comment by MaxKellermann** [comment]
> Yeah, the `kill` flag _can_ be the thing to check, but I had doubts about it actually working because of this code path:

This code path is bad IMO. What's the point of suppressing the notify? A micro-optimization which shaves off a few nanoseconds from the shutdown process? That's pointless.

A notify on a `condition_variable` nobody is waiting on is a no-op; not even a system call is performed (at least on Linux/glibc). See https://github.com/bminor/glibc/blob/master/nptl/pthread_cond_signal.c

> Still don't think so unless I read the above logic incorrectly. That's why I was suggesting a separate `woke` state, since `WakeUp` can be called when `killed` is false and does nothing to try to change that state in order to prevent the target thread from waiting.

What's the point of this `woke` state and what should the thread do then?

For the sake of omitting the `wait()` call, that flag is pointless - because omitting the `wait()` call will simply turn this into a busy loop, defeating the purpose of the `condition_variable`.

Why would you wake the thread? The code defines two conditions already:
https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/Queue.h#L160
1. reader/writer got killed (this was my example)
2. there is new data

If any of this is true, then the thread should wake up and act (terminate or handle the new data).
Once all data is consumed and the queue has become empty again, go back to sleep until the next notify is received.

Bottom line: Zeek alreay has this `woke` state, and adding another flag will only add complexity and make the code more fragile and less elegant and less readable.

What kind of race condition still exists in your opinion? (Other than the unprotected `killed` flags, which is out of this PR's scope)

**Comment by jsiwek** [comment]
> This code path is bad IMO. What's the point of suppressing the notify? A micro-optimization which shaves off a few nanoseconds from the shutdown process? That's pointless.

Agree, but also now thinking that `WakeUp()` doesn't serve a purpose in the general sense -- there's already a notify associated with changing the `killed` flag and also on a new message becoming available.

> Bottom line: Zeek alreay has this `woke` state, and adding another flag will only add complexity and make the code more fragile and less elegant and less readable.

Agree, can drop that idea -- I was confused by the above (miss)use of `WakeUp()` and was convinced there was a missing termination/wakeup condition.  Agree that the two existing wakeup conditions are valid/sufficient.

> What kind of race condition still exists in your opinion? (Other than the unprotected `killed` flags, which is out of this PR's scope)

Just the unprotected `killed` flag now, but by this point, I see the distinction and good to proceed with this PR (though the way that flag is dissociated maybe contributed to my confusion).  Anyways: thanks.

**Comment by jsiwek** [comment]
> Agree, but also now thinking that `WakeUp()` doesn't serve a purpose in the general sense

Just re-read that and found it could parse badly -- I was just affirming that 1 particular `WakeUp()` call I had quoted earlier as being weird:

https://github.com/zeek/zeek/blob/36557f3086c95f079aa3e69c26a34adeb493c73f/src/threading/MsgThread.cc#L245-L246

The other existing call to `WakeUp()` does make sense to me.
